### PR TITLE
ST-384 Spike

### PR DESCRIPTION
**Pull request by SeanArmstrong**
- [ ] ~Unit tests passing~
- [ ] ~Coverage reached~
- [ ] ~Acceptance tests passing~
- [ ] ~Reviewed~

### CODE REVIEW

File: app/controllers/income/UpdateIncomeNextYearController.scala [hunk 1]
```diff
@@ -22,14 +22,24 @@ import uk.gov.hmrc.play.partials.FormPartialRetriever
...
-import uk.gov.hmrc.tai.service.JourneyCacheService
-import uk.gov.hmrc.tai.util.JourneyCacheConstants
+import uk.gov.hmrc.tai.service.{JourneyCacheService, UpdateNextYearsIncomeService}
+import uk.gov.hmrc.tai.util.constants.journeyCache.UpdateNextYearsIncomeConstants
 
 trait UpdateIncomeNextYearController extends TaiBaseController {
 
-  def journeyCacheService: JourneyCacheService
+  def start(employmentId: Int): Action[AnyContent] = {
+//    implicit person =>
```
Comment (amaalali) [inline app/controllers/income/UpdateIncomeNextYearController.scala]: We can remove this

File: app/uk/gov/hmrc/tai/connectors/JourneyCacheConnector.scala [hunk 2]
```diff
@@ -51,7 +51,9 @@ trait JourneyCacheConnector {
   }
 
   def cache(journeyName: String, data: Map[String,String])(implicit hc: HeaderCarrier): Future[Map[String,String]] = {
-    httpHandler.postToApi(cacheUrl(journeyName), data).map(_.json.as[Map[String,String]])
+    httpHandler.postToApi(
+      cacheUrl(journeyName), data
+    ).map(_.json.as[Map[String,String]])
```
Comment (amaalali) [inline app/uk/gov/hmrc/tai/connectors/JourneyCacheConnector.scala]: This is where we can make this API more flexible by allowing `cache` to take a type param here and that will specify the return type

File: app/uk/gov/hmrc/tai/service/UpdateNextYearsIncomeService.scala [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+import uk.gov.hmrc.domain.Nino
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.http.logging.MdcLoggingExecutionContext._
+import uk.gov.hmrc.tai.model.TaxYear
+import uk.gov.hmrc.tai.util.constants.journeyCache.UpdateNextYearsIncomeConstants
+
+import scala.concurrent.Future
+
+final case class UpdateNextYearsIncomeCacheModel(employmentName: String, employmentId: Int, currentValue: BigDecimal, newValue: Option[Int] = None) {
```
Comment (amaalali) [inline app/uk/gov/hmrc/tai/service/UpdateNextYearsIncomeService.scala]: We should move this over to its own package

File: test/uk/gov/hmrc/tai/service/UpdateNextYearsIncomeServiceSpec.scala [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        val nino = generateNino
+
+        when(updateNextYearsIncomeService.employmentService.employment(Matchers.eq(nino), Matchers.eq(employmentId))(any()))
+          .thenReturn(Future.successful(Some(employment(employmentName))))
+
+        when(updateNextYearsIncomeService.taxAccountService.taxCodeIncomeForEmployment(
+          Matchers.eq(nino), Matchers.eq(TaxYear()), Matchers.eq(employmentId))(any())
+        ).thenReturn(Future.successful(None))
+
+        val result: RuntimeException = the[RuntimeException] thrownBy Await.result(updateNextYearsIncomeService.setup(employmentId, nino), 5 seconds)
```
Comment (amaalali) [inline test/uk/gov/hmrc/tai/service/UpdateNextYearsIncomeServiceSpec.scala]: I _think_ we need to intercept this still or else it pollutes the output of the tests

File: app/uk/gov/hmrc/tai/service/UpdateNextYearsIncomeService.scala [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+      employmentOption <- employmentFuture
+    } yield (taxCodeIncomeOption, employmentOption) match {
+      case (Some(taxCodeIncome), Some(employment)) => {
+        val model = UpdateNextYearsIncomeCacheModel(employment.name, employmentId, taxCodeIncome.amount)
+
+        journeyCacheService.cache(model.toCacheMap)
+
+        model
+      }
+      case _ => throw new RuntimeException("[UpdateNextYearsIncomeService] Could not set up next years estimated income journey")
```
Comment (amaalali) [inline app/uk/gov/hmrc/tai/service/UpdateNextYearsIncomeService.scala]: I'm not a big fan of this throwing error way of handling cases where there is no data. We can probably discuss with design on if we can show a more meaningful page than the standard error page

Comment (SeanArmstrong) [reply app/uk/gov/hmrc/tai/service/UpdateNextYearsIncomeService.scala]: This is true, we dont want to block the rest of the work based on this though. Perhaps a TODO would be appropriate

File: test/uk/gov/hmrc/tai/service/UpdateNextYearsIncomeServiceSpec.scala (hunk 4 above)
Comment (SeanArmstrong) [reply test/uk/gov/hmrc/tai/service/UpdateNextYearsIncomeServiceSpec.scala]: Can you elaborate on this

Comment (SeanArmstrong) [reply test/uk/gov/hmrc/tai/service/UpdateNextYearsIncomeServiceSpec.scala]: Im not seeing any bad output from the tests
```
[info] Configuring Artifactory... Host: not set. User: not set. Password defined: false
[info] Compiling 2 Scala sources to /Users/seanarmstrong/Development/tai-frontend/target/scala-2.11/classes...
[info] UpdateNextYearsIncomeServiceSpec:
[info] setup
[info]   must initialize the journey cache and return the cache model
[info]   - when an taxCodeIncome and Employment is returned
[info]   must throw a runtime exception
[info]   - when could not retrieve a TaxCodeIncome
[info]   - when could not retrieve a Employment
[info] get
[info]   must retrieve the cached values and returns the cache model
[info]   - when journey values exist without a new amount in the cache
[info]   - when journey values exist with a new amount in the cache
[info]   must setup the cache
[info]   - when journey values do not exist in the cache
[info] setNewAmount
[info]   must update the cache with the new amount
[info]   - when journey values are successfully retrieved from the cache
[info] ScalaCheck
[info] Passed: Total 0, Failed 0, Errors 0, Passed 0
[info] Warning: Unknown ScalaCheck args provided: -o -u target/test-reports -h target/test-reports/html-report -o -u target/test-reports -h target/test-reports/html-report
[info] ScalaTest
[info] Run completed in 3 seconds, 377 milliseconds.
[info] Total number of tests run: 7
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 7, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

**Pull request merged**